### PR TITLE
Serve unified controller api related to java

### LINKED ISSUE #22858: [serve]The interaction between Java API and Unified Controller API
In a complete Java user case, the cross language interaction between Java API and Python controller will be involved, such as starting a serve, deploying a deployment, querying the deployment status, and so on. The proposal of [Serve Unified Controller API](https://docs.google.com/document/d/1Yb9eZIQ4cZWypWZj4bK__7nDXWgthoj52QRg8FOFmNY/edit#heading=h.mttk25m16f6c) can solve this cross language problem. 

This issue lists the parts where the Java API interacts with the Controller API, as well as the related modification schemes.

### deploy
parameters:
| parameter name | parameter type |
| :----: | :----: |
| name | str |
| deployment_config_proto_bytes | bytes |
| replica_config | ReplicaConfig |
| version | Optional[str] |
| prev_version | Optional[str] |
| route_prefix | Optional[str] |
| deployer_job_id | "ray._raylet.JobID" |

* Merge `version`, `prev_version` into `DeploymentConfig`.
* Define `ReplicaConfig` protobuf.

### get_http_proxies
return: `Dict[NodeId, ActorHandle]`
* Change to the `ActorList` protobuf which contains an actor name list.

### get_deployment_info
return: `Tuple[DeploymentInfo, str]`
* Define a protobuf like `DeploymentRoute` composed of the `DeploymentInfo` and the `route_prefix`.

### list_deployments
return: `Dict[str, Tuple[DeploymentInfo, str]]`
* Define `DeploymentRouteList` protobuf containing the `DeploymentRoute` list.

### get_deployment_statuses
return: `Dict[str, DeploymentStatusInfo]`
* Define `DeploymentStatusInfo` protobuf.
* Define `DeploymentStatusInfoList` containing the `DeploymentStatusInfo` list.

### PR DESCRIPTION
**Pull request by liuyang-my**
## Why are these changes needed?

This PR submitted changes related to the Serve Unified controller API(Xlang Java).

## Related issue number

#22858

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/api.py [hunk 1]
```diff
@@ -210,7 +214,7 @@ def _wait_for_deployments_shutdown(self, timeout_s: int = 60):
         """
         start = time.time()
         while time.time() - start < timeout_s:
-            statuses = ray.get(self._controller.get_deployment_statuses.remote())
+            statuses = get_deployment_statuses()
```
Comment (simon-mo) [inline python/ray/serve/api.py:217]: self.get_deployment_statuses?

File: python/ray/serve/config.py [hunk 2]
```diff
@@ -129,6 +130,10 @@ class DeploymentConfig(BaseModel):
     # the deploymnent use.
     deployment_language: Any = DeploymentLanguage.PYTHON
 
+    name: str
+    version: Optional[str]
```
Comment (shrekris-anyscale) [inline python/ray/serve/config.py:134]: I think we can remove `prev_version`. Was there a verdict on `version`? @simon-mo @edoakes

File: python/ray/serve/config.py [hunk 3]
```diff
@@ -285,6 +297,40 @@ def _validate(self):
...
+            use_integers_for_enums=True,
+        )
+        if "init_args" in data:
+            if data["init_args"] != "":
+                data["init_args"] = pickle.loads(proto.init_args)
+            else:
+                data["init_args"] = None
+        if "init_kwargs" in data:
+            if data["init_kwargs"] != "":
+                data["init_kwargs"] = pickle.loads(proto.init_args)
```
Comment (shrekris-anyscale) [inline python/ray/serve/config.py:315]: This should be `proto.init_kwargs`.

File: python/ray/serve/controller.py [hunk 4]
```diff
@@ -296,6 +301,11 @@ def deploy(
         deployment_config = DeploymentConfig.from_proto_bytes(
             deployment_config_proto_bytes
         )
+        name = deployment_config.name
+        version = deployment_config.version
```
Comment (shrekris-anyscale) [inline python/ray/serve/controller.py:305]: See other comment about `version` and `prev_version`.

File: python/ray/serve/controller.py [hunk 5]
```diff
@@ -407,5 +419,34 @@ def list_deployments(
...
+
+        Args:
+            include_deleted(bool): Whether to include information about
+                deployments that have been deleted.
+
+        Returns:
+            DeploymentRoute's protobuf
+
+        Raises:
+            KeyError if the deployment doesn't exist.
```
Comment (shrekris-anyscale) [inline python/ray/serve/controller.py:435]: Which deployment does this refer to?

File: src/ray/protobuf/serve.proto [hunk 6]
```diff
@@ -86,6 +86,12 @@ message DeploymentConfig {
 
   // The deployment's autoscaling configuration.
   AutoscalingConfig autoscaling_config = 10;
+
+  string name = 11;
+
+  string version = 12;
```
Comment (shrekris-anyscale) [inline src/ray/protobuf/serve.proto:92]: See previous comment about `version` and `prev_version`.

File: src/ray/protobuf/serve.proto [hunk 7]
```diff
@@ -139,3 +149,45 @@ message DeploymentVersion {
...
+  DeploymentConfig deployment_config = 1;
+  ReplicaConfig replica_config = 2;
+  int32 start_time_ms = 3;
+  string actor_name = 4;
+  bytes serialized_deployment_def = 5;
+  string version = 6;
+  int32 end_time_ms = 7;
+}
+
+message DeploymentRoute {
```
Comment (shrekris-anyscale) [inline src/ray/protobuf/serve.proto:170]: Is there a reason we keep this separate from `DeploymentInfo`? Can we put `route` in `DeploymentInfo`? And can we change `DeploymentRouteList` into `DeploymentInfoList`? cc @simon-mo @edoakes

**Review by shrekris-anyscale — DISMISSED** [review]
Great work on the unified API! I left some feedback. Please let me know if you have any questions or comments.

File: python/ray/serve/api.py (hunk 1 above)
Comment (liuyang-my) [reply python/ray/serve/api.py]: Yes. It is fixed now.

File: python/ray/serve/config.py (hunk 3 above)
Comment (liuyang-my) [reply python/ray/serve/config.py]: It is fixed now.

File: python/ray/serve/controller.py (hunk 5 above)
Comment (liuyang-my) [reply python/ray/serve/controller.py]: The `RaiseIt` comment is deleted now.

File: python/ray/serve/config.py (hunk 2 above)
Comment (liuyang-my) [reply python/ray/serve/config.py]: In order to quickly support Java user scenarios in this PR, no `version` and `prev_version` related changes are made. Can you make relevant adjustments later? @simon-mo

File: src/ray/protobuf/serve.proto (hunk 7 above)
Comment (liuyang-my) [reply src/ray/protobuf/serve.proto:171]: In order not to destroy the existing `DeploymentInfo` structure, the `DeploymentInfo` and `route` are encapsulated in the `DeploymentRoute`. Can the integration of deploymentinfo and route be evaluated later?
 @simon-mo

File: src/ray/protobuf/serve.proto [hunk 8]
```diff
@@ -131,6 +135,11 @@ message EndpointSet {
   map<string, EndpointInfo> endpoints = 1;
 }
 
+message ActorNameList {
```
Comment (simon-mo) [inline src/ray/protobuf/serve.proto:138]: document why do we need this (instead of actor handles)

File: src/ray/protobuf/serve.proto [hunk 9]
```diff
@@ -131,6 +135,11 @@ message EndpointSet {
...
+  repeated string names = 1;
+}
+
+// It is deprecated. Use ActorNameList instead.
 message ActorSet {
   repeated string names = 1;
 }
```
Comment (simon-mo) [inline src/ray/protobuf/serve.proto:145]: can we just remove ActorSet now?

File: src/ray/protobuf/serve.proto (hunk 7 above)
Comment (simon-mo) [reply src/ray/protobuf/serve.proto:171]: I think looks fine for now!

File: python/ray/serve/controller.py [hunk 10]
```diff
@@ -418,5 +435,43 @@ def list_deployments(
...
+            deployment_info_proto = deployment_info.to_proto()
+            deployment_info_proto.name = deployment_name
+            deployment_route_list.deployment_routes.append(
+                DeploymentRoute(
+                    deployment_info=deployment_info_proto, route=route_prefix
+                )
+            )
+        return deployment_route_list.SerializeToString()
+
+    def get_deployment_statuses(self) -> bytes:
```
Comment (simon-mo) [inline python/ray/serve/controller.py:464]: docstring

File: python/ray/serve/controller.py [hunk 11]
```diff
@@ -418,5 +435,43 @@ def list_deployments(
...
-        return self.deployment_state_manager.get_deployment_statuses()
+    def list_deployments(self, include_deleted: Optional[bool] = False) -> bytes:
+        """Gets the current information about all deployments.
+
+        Args:
+            include_deleted(bool): Whether to include information about
+                deployments that have been deleted.
+
+        Returns:
+            DeploymentRoute's protobuf
```
Comment (simon-mo) [inline python/ray/serve/controller.py:446]:
```suggestion
            DeploymentRouteList's protobuf serialized bytes
```

**Review by simon-mo — APPROVED** [review]
Looks good overall! Few minor comments.

File: src/ray/protobuf/serve.proto (hunk 9 above)
Comment (liuyang-my) [reply src/ray/protobuf/serve.proto:149]: Java LongPollClient depends on it now. I will replace it to ActorNameList in next Java API's PR.

File: src/ray/protobuf/serve.proto (hunk 8 above)
Comment (liuyang-my) [reply src/ray/protobuf/serve.proto:142]: Document is added.

File: python/ray/serve/controller.py (hunk 10 above)
Comment (liuyang-my) [reply python/ray/serve/controller.py:464]: Docstring is added.

File: python/ray/serve/controller.py (hunk 11 above)
Comment (liuyang-my) [reply python/ray/serve/controller.py]: It has been revised according to the suggested change.

**Review by jovany-wang — APPROVED** [review]
@simon-mo Could you help take a look at the CI result. I'm not sure if it's related.

**Pull request merged**